design/safe-disruption-of-nodes-with-osds.md

```diff
@@ -0,0 +1,30 @@
+# Safe disruption of nodes with osds
+
+
+External operators and agents that provision,fence, or upgrade nodes should be able to use the kubernetes api to know how not to disrupt storage in the course of their operation.
+
+External agents should only bring down one failure domain at once.
+
+
+## Assumptions:
+- The failure domain is node/host(or higher)
+- To preserve availability, it is enough to ensure that only one host-that-has osds is down at once. 
+- It is fine for the operator to constantly ping the cluster for it’s status.
+
+## Proposed changes to `rook-ceph`:
+
+- The nodes with osds can be labeled with `ceph.rook.io/has-osds=true`
+- The nodes with osds can be annotated with the list of clusters that have osds on it.
+    - The key can be `ceph.rook.io/tenant-clusters`
+- The cluster CRD status can be updated with `CephCluster.Status.ClusterHealth` and `CephCluster.Status.ClusterHealthLastUpdated`.
+
+## Example agent behaviour:
+
+
+- The agent should only carry out disruptive actions on a node with `ceph.rook.io/has-osds=true` **only if** these conditions are satisfied:
```

Comment:
If each agent has to implement this, the upgrade reliability will be dependent on the agent. Is there a more generic mechanism Rook can trigger to make this a general solution? For example, can you add details about the limitations of the pod disruption budgets?

Reply:
The agent referenced here is any agent that would cause node level downtime - like a kernel upgrade. Since ceph osds are tied to the node, we need to ensure that for each cluster, a minimum of the nodes that it has osds on is up at all times.

Reply:
I'll add details about why PodDisruptionBudgets don't meet the requirements.

Reply:
It's in the design now.